File: apps/api/ids/ids.go
// Package ids encodes and decodes Tracky's typed global identifiers.
//
// Every addressable resource is referred to by a self-describing ID of the form
//
//	[kind]-[provider]-[native]
//
// where kind is a single-character entity tag (s, r, t, h) and native is the
// provider's own GTFS identifier. Operators are a degenerate case with no
// native id: o-[provider].
//
// The `-` is the structural separator. Within a single segment (provider or
// native), the `~` character is permitted as a word-break — useful for
// multi-word provider names that don't fit a single token.
//
// Examples:
//   - s-amtrak-CHI                  stop
//   - r-amtrak-40751                route
//   - t-amtrak-251208               trip
//   - h-amtrak-NYC                  hub (meta-station)
//   - o-amtrak                      operator / provider
//   - s-metra~electric-FOO          multi-word provider
//   - t-brightline-service~A~v2     tildes inside the native id
package ids

import (
	"errors"
	"fmt"
	"strings"
)

type Kind string

const (
	KindStop     Kind = "s"
	KindRoute    Kind = "r"
	KindTrip     Kind = "t"
	KindHub      Kind = "h"
	KindOperator Kind = "o"
)

var knownKinds = map[Kind]bool{
	KindStop: true, KindRoute: true, KindTrip: true, KindHub: true, KindOperator: true,
}

// ID is the decoded form of a typed global identifier.
// Native is empty when Kind == KindOperator.
type ID struct {
	Kind     Kind
	Provider string
	Native   string
}

var (
	ErrEmpty          = errors.New("ids: empty input")
	ErrMissingDash    = errors.New("ids: missing '-' separator")
	ErrUnknownKind    = errors.New("ids: unknown kind")
	ErrEmptyProvider  = errors.New("ids: empty provider")
	ErrEmptyNative    = errors.New("ids: empty native id")
	ErrOperatorNative = errors.New("ids: operator id must not have a native segment")
	ErrProviderDash   = errors.New("ids: provider must not contain '-' (use '~' for multi-word providers)")
)

// Encode builds a global ID from its parts. Returns an error if any part
// violates the format (empty provider, '-' in provider, etc.).
func Encode(kind Kind, provider, native string) (string, error) {
	if !knownKinds[kind] {
		return "", fmt.Errorf("%w: %q", ErrUnknownKind, kind)
	}
	if provider == "" {
		return "", ErrEmptyProvider
	}
	if strings.ContainsRune(provider, '-') {
		return "", fmt.Errorf("%w: %q", ErrProviderDash, provider)
	}
	if kind == KindOperator {
		if native != "" {
			return "", ErrOperatorNative
		}
		return string(kind) + "-" + provider, nil
	}
	if native == "" {
		return "", ErrEmptyNative
	}
	return string(kind) + "-" + provider + "-" + native, nil
}

// MustEncode is Encode that panics on error. For tests and constants only.
func MustEncode(kind Kind, provider, native string) string {
	s, err := Encode(kind, provider, native)
	if err != nil {
		panic(err)
	}
	return s
}

// Decode parses a global ID. Operator IDs (o-foo) are accepted with an empty
// Native field; all other kinds require a non-empty native after the second '-'.
func Decode(s string) (ID, error) {
	if s == "" {
		return ID{}, ErrEmpty
	}
	// First dash splits kind from the rest.
	kindStr, rest, ok := strings.Cut(s, "-")
	if !ok || kindStr == "" {
		return ID{}, ErrMissingDash
	}
	kind := Kind(kindStr)
	if !knownKinds[kind] {
		return ID{}, fmt.Errorf("%w: %q", ErrUnknownKind, kind)
	}
	if kind == KindOperator {
		if strings.ContainsRune(rest, '-') {
			return ID{}, ErrOperatorNative
		}
		if rest == "" {
			return ID{}, ErrEmptyProvider
		}
		return ID{Kind: kind, Provider: rest}, nil
	}
	// Second dash splits provider from native. Native may itself contain '-'
	// (we cut on the first one only) so e.g. native='NY-PENN' is fine.
	provider, native, ok := strings.Cut(rest, "-")
	if !ok {
		return ID{}, ErrMissingDash
	}
	if provider == "" {
		return ID{}, ErrEmptyProvider
	}
	if native == "" {
		return ID{}, ErrEmptyNative
	}
	return ID{Kind: kind, Provider: provider, Native: native}, nil
}

// DecodeKind parses s and asserts its kind matches want. Convenience for
// handlers that already know which kind they expect from the route.
func DecodeKind(s string, want Kind) (ID, error) {
	id, err := Decode(s)
	if err != nil {
		return ID{}, err
	}
	if id.Kind != want {
		return ID{}, fmt.Errorf("ids: expected kind %q, got %q", want, id.Kind)
	}
	return id, nil
}

// String returns the encoded form of id, or "" if id is invalid.
func (id ID) String() string {
	s, err := Encode(id.Kind, id.Provider, id.Native)
	if err != nil {
		return ""
	}
	return s
}

File: apps/mobile/utils/ids.ts
/**
 * Encodes and decodes Tracky's typed global identifiers.
 *
 * Format: `[kind]-[provider]-[native]`
 *   - s-amtrak-CHI                  stop
 *   - r-amtrak-40751                route
 *   - t-amtrak-251208               trip
 *   - h-amtrak-NYC                  hub (meta-station)
 *   - o-amtrak                      operator / provider  (no native segment)
 *   - s-metra~electric-FOO          multi-word provider
 *   - t-brightline-service~A~v2     tildes inside the native id
 *
 * The `-` is the structural separator. Within a single segment (provider or
 * native), the `~` character is permitted as a word-break — useful for
 * multi-word provider names that don't fit a single token.
 *
 * Mirrors apps/api/ids/ids.go — keep the two in sync.
 */

export type IdKind = 's' | 'r' | 't' | 'h' | 'o';

export interface ParsedId {
  kind: IdKind;
  provider: string;
  /** Empty when kind === 'o'. */
  native: string;
}

const KNOWN_KINDS: ReadonlySet<string> = new Set(['s', 'r', 't', 'h', 'o']);

export class IdFormatError extends Error {
  constructor(message: string, public readonly input: string) {
    super(message);
    this.name = 'IdFormatError';
  }
}

export function encodeId(kind: IdKind, provider: string, native: string): string {
  if (!KNOWN_KINDS.has(kind)) {
    throw new IdFormatError(`unknown kind ${JSON.stringify(kind)}`, kind);
  }
  if (!provider) throw new IdFormatError('empty provider', '');
  if (provider.includes('-')) {
    throw new IdFormatError(
      `provider ${JSON.stringify(provider)} must not contain '-' (use '~' for multi-word providers)`,
      provider,
    );
  }
  if (kind === 'o') {
    if (native) throw new IdFormatError('operator id must not have a native segment', native);
    return `o-${provider}`;
  }
  if (!native) throw new IdFormatError('empty native id', '');
  return `${kind}-${provider}-${native}`;
}

/**
 * Parse a global ID. Throws IdFormatError on malformed input.
 * Use tryParseId() when you want a nullable return instead.
 */
export function parseId(s: string): ParsedId {
  if (!s) throw new IdFormatError('empty input', s);
  const dash = s.indexOf('-');
  if (dash <= 0) throw new IdFormatError("missing '-' separator", s);
  const kindStr = s.slice(0, dash);
  if (!KNOWN_KINDS.has(kindStr)) {
    throw new IdFormatError(`unknown kind ${JSON.stringify(kindStr)}`, s);
  }
  const kind = kindStr as IdKind;
  const rest = s.slice(dash + 1);
  if (kind === 'o') {
    if (rest.includes('-')) throw new IdFormatError('operator id must not have a native segment', s);
    if (!rest) throw new IdFormatError('empty provider', s);
    return { kind, provider: rest, native: '' };
  }
  // Native may itself contain '-' — we cut on the first one only, so e.g.
  // native='NY-PENN' resolves correctly.
  const second = rest.indexOf('-');
  if (second < 0) throw new IdFormatError("missing '-' separator before native id", s);
  const provider = rest.slice(0, second);
  const native = rest.slice(second + 1);
  if (!provider) throw new IdFormatError('empty provider', s);
  if (!native) throw new IdFormatError('empty native id', s);
  return { kind, provider, native };
}

export function tryParseId(s: string): ParsedId | null {
  try {
    return parseId(s);
  } catch {
    return null;
  }
}

/** Parse s and assert its kind matches want. Returns null on any failure. */
export function tryParseKindedId(s: string, want: IdKind): ParsedId | null {
  const parsed = tryParseId(s);
  if (!parsed || parsed.kind !== want) return null;
  return parsed;
}

/** Convenience: extract the provider segment without allocating a ParsedId. */
export function providerOf(globalId: string): string | null {
  return tryParseId(globalId)?.provider ?? null;
}

/**
 * Composite identity for a single run (trip × run date). Use as a React key
 * or dedup key in any list that may contain multiple runs of the same trip
 * (e.g. yesterday's Amtrak 21 still en route while today's just departed).
 *
 * Format: `<tripId>@<day-of-month>`. Day-of-month is sufficient to
 * disambiguate concurrent runs — no real-world trip lasts long enough to
 * collide with itself a month later (longest Amtrak end-to-end is ~65h).
 * `@` is deliberately distinct from `-` (typed-id structural separator) and
 * `~` (typed-id word-break) so a run key can never be misparsed as a global
 * id. This is a purely client-side convention — see apps/api/spec/realtime.go
 * for the server-side helper.
 *
 * `runDate` may be a YYYY-MM-DD string or a Date / epoch-millis.
 */
export function runKey(run: {
  tripId: string;
  runDate?: string | number | Date | null;
}): string {
  const day = dayOfMonth(run.runDate);
  return day == null ? run.tripId : `${run.tripId}@${day}`;
}

function dayOfMonth(d: string | number | Date | null | undefined): number | null {
  if (d == null) return null;
  // Strings come in two flavors from the wire / saved data:
  //   - YYYY-MM-DD              (Go's Format("2006-01-02"))
  //   - RFC3339 / ISO 8601      (default time.Time JSON encoding)
  // Date parsing handles both. UTC-day extraction is correct because the
  // backend stamps run_date as UTC midnight of the service date.
  const date = d instanceof Date ? d : new Date(d);
  return Number.isFinite(date.getTime()) ? date.getUTCDate() : null;
}
